streams/shared/src/main/scala/zio/stream/ZChannel.scala

```diff
@@ -1940,111 +1940,120 @@ object ZChannel {
-                           (fiberId: FiberId) =>
-                             pull.flatMap { channel =>
-                               val canceler = Promise.unsafe.make[Nothing, Unit](fiberId)(Unsafe.unsafe)
-                               val latch    = Promise.unsafe.make[Nothing, Unit](fiberId)(Unsafe.unsafe)
-                               for {
-                                 size <- cancelers.size
-                                 _    <- ZIO.when(size >= n)(cancelers.take.flatMap(_.succeed(())))
-                                 _    <- cancelers.offer(canceler)
-                                 raceIOs =
-                                   ZIO.scopedWith { scope =>
-                                     (incoming >>> channel)
-                                       .toPullInAlt(scope)
-                                       .flatMap(evaluatePull(_).race(canceler.await.interruptible))
-                                   }
-                                 _ <- permits
-                                        .withPermit(latch.succeed(()) *> raceIOs)
-                                        .interruptible
-                                        .forkIn(childScope)
-                                 _ <- latch.await
-                               } yield ()
-                             }
-                       }
-        _ <- pullStrategy(fiberId).forever
-               .catchAllCause(cause =>
-                 cause.failureOrCause match {
-                   case Left(_: Left[OutErr, OutDone]) => outgoing.offer(Exit.failCause(cause))
-                   case Left(x: Right[OutErr, OutDone]) =>
-                     permits.withPermits(n.toLong)(ZIO.unit).interruptible *>
-                       lastDone.get.flatMap {
-                         case Some(lastDone) => outgoing.offer(Exit.fail(Right(f(lastDone, x.value))))
-                         case None           => outgoing.offer(Exit.fail(Right(x.value)))
-                       }
-                   case Right(cause) => outgoing.offer(Exit.failCause(cause.map(Left(_))))
-                 }
-               )
-               .raceFirst(awaitErrorSignal(childScope, fiberId)(errorSignal))
-               .forkIn(scope)
-      } yield {
-        lazy val consumer: ZChannel[Env, Any, Any, Any, OutErr, OutElem, OutDone] =
-          unwrap[Env, Any, Any, Any, OutErr, OutElem, OutDone] {
-            outgoing.take.flatten.foldCause(
-              cause =>
+      ZIO.fiberIdWith { fiberId =>
+        ZIO.suspendSucceed {
+          val input          = SingleProducerAsyncInput.unsafe.make[InErr, InElem, InDone](fiberId)
+          val incoming       = ZChannel.fromInput(input)
+          val n0             = n.toLong
+          val bufferSize0    = bufferSize
+          val mergeStrategy0 = mergeStrategy
+          val outgoing =
+            Queue.unsafe.bounded[ZIO[Env, Either[OutErr, OutDone], OutElem]](bufferSize0, fiberId)(Unsafe.unsafe)
```

Comment:
by the by, I think this could be
```suggestion
            Queue.unsafe.bounded[Exit[Either[OutErr, OutDone], OutElem]](bufferSize0, fiberId)(Unsafe.unsafe)
```
then `consumer` can avoid the `flatten` and just pattern match:
```
outgoing.take.map {
  case s: Success[OutElem] => ZChannel.write(s.value) *> write
  case f: Failure[Either[OutErr, OutDone]] => ...
}
```

Reply:
We could go even further and use a specialized data structure which could only be something like:
```
sealed trait Result[V, E, F]
object Result {
  final case class Value[V](v: V) extends Result[V, Nothing, Nothing]
  final case class Error[E](e: E) extends Result[Nothing, E, Nothing]
  final case class Fatal[E](e: E) extends Result[Nothing, Nothing, E]
}
```

That'd be more lightweight than `Exit[Either[OutErr, OutDone], OutElem]`.

I'll give it a try

Reply:
Nice one 👌 Another iteration on this that was exploited previously is to avoid wrapping the `OutElem` values, turning the queue elements into `OutElem | Result`.

Reply:
Does `OutElem | Result` work with Scala2? 🤔